src/neural/cuda/winograd_helper.inc

```diff
@@ -0,0 +1,387 @@
+  matrixMul_gpu_serial<T, 6, 6, 3>(transformed_filter, temp_filter, Gt);
+}
+
+template <typename T>
+__device__ __forceinline__ void InputTransform4x4(T* transformedInput,
+                                                  const T* input) {
+  // transform applied to input tile (of size 4x4)
+  const T Bt[6 * 6] = {4, 0, -5, 0,  1, 0, 0, -4, -4, 1,  1, 0,
+                       0, 4, -4, -1, 1, 0, 0, -2, -1, 2,  1, 0,
+                       0, 2, -1, -2, 1, 0, 0, 4,  0,  -5, 0, 1};
+
+  const T B[6 * 6] = {4,  0,  0,  0,  0,  0, 0, -4, 4,  -2, 2,  4,
+                      -5, -4, -4, -1, -1, 0, 0, 1,  -1, 2,  -2, -5,
+                      1,  1,  1,  1,  1,  0, 0, 0,  0,  0,  0,  1};
+
+  T tempIp1[6 * 6];
+  matrixMul_gpu_serial<T, 6, 6, 6>(tempIp1, Bt, input);
+  matrixMul_gpu_serial<T, 6, 6, 6>(transformedInput, tempIp1, B);
+}
+
+template <typename T>
+__device__ __forceinline__ void OutputTransform4x4(
+    T* output, const T* transformedOutput) {
+  // transform applied to result
+  const T At[4 * 6] = {1, 1, 1, 1, 1, 0, 0, 1, -1, 2, -2, 0,
+                       0, 1, 1, 4, 4, 0, 0, 1, -1, 8, -8, 1};
+
+  const T A[6 * 4] = {1, 0, 0, 0, 1, 1,  1, 1,  1, -1, 1, -1,
+                      1, 2, 4, 8, 1, -2, 4, -8, 0, 0,  0, 1};
+
+  T tempOp[4 * 6];
+  matrixMul_gpu_serial<T, 4, 6, 6>(tempOp, At, transformedOutput);
+  matrixMul_gpu_serial<T, 4, 4, 6>(output, tempOp, A);
+}
+
+#define FILTER_IDX_NCHW(k, c, h, w) ((k)*C * S * R + (c)*S * R + (h)*R + w)
+template <typename T>
+__global__ void filterTransform_kernel(int K, int C, int elements,
+                                       T* transformed_filter, const T* filter) {
+  int tid = blockIdx.x * blockDim.x + threadIdx.x;
+  if (tid >= elements) return;
+
+  constexpr int S = 3;
+  constexpr int R = 3;
+
+  int c = tid % C;
+  int k = tid / C;
+
+  T filter_tile[3][3];
+  T transformed_tile[6][6];
+
+  // read input from memory
+  for (int s = 0; s < S; s++)
+    for (int r = 0; r < R; r++) {
+      filter_tile[s][r] = filter[FILTER_IDX_NCHW(k, c, s, r)];
+    }
+
+  // transform it
+  FilterTransform4x4(&(transformed_tile[0][0]), &(filter_tile[0][0]));
+
+  // write to output (output is in HWCK layout)
+  for (int i = 0; i < 6; i++)
+    for (int j = 0; j < 6; j++) {
+      transformed_filter[i * 6 * C * K + j * C * K + c * K + k] =
+          transformed_tile[i][j];
+    }
+}
+
+
+// index in intermediate/temp tensor
+// W, H == 6 here! (6x6 transformed blocks)
+// N also includes part of dimension (2x2)
+#define GemmN (N * 4)
+#define INDEX_NCHW(n, c, h, w) ((n)*C * 8 * 8 + (c)*8 * 8 + (h)*8 + w)
+#define TEMP_INDEX_HWNC(h, w, n, c) \
+  ((h)*6 * GemmN * C + (w)*GemmN * C + (n)*C + c)
+
+// 'C' threads per block
+// 'N' blocks
+// every thread transforms an entire board/plane (8x8 elements)
+// - producing 4 x 6x6 elements
+template <typename T>
+__global__ void InputTransform_kernel(int N, int C, const T* input, T* output) {
+  constexpr int H = 8, W = 8;
```

Comment:
nvcc warns that H and W are not used.

Reply:
Thanks for catching. Removed them. There are other assumptions based on W being 8, so no point in defining as constant with the hope that the kernel will still work if someone changes it to something else.